### DISCUSSION TITLE
Reports - Help w/ Geocoder MIssing Error

### DISCUSSION THREAD

**Discussion by tonydm** [netbox-community/netbox#8513] [Q&A]
Hello, I'm getting familiar with NetBox and copied some Report scripts from https://github.com/netbox-community/reports/tree/master/reports.  I copied the site_address.py script to the reports dir.  This may or may not be a report that I'd be interested in in production.  But to determine that, I would like to run it w/o errors.  I ran `pip install geocoder` successfully.  I can't seem to find the requirements.txt or local_requirements.txt file(s).  I'm not a Python coder so really not familiar with working with it. 
 Any feedback or suggestions to resolve would be appreciated.

Thank you

HK

## Server Error
A module import error occurred during this request. Common causes include the following:

**Missing required packages** - This installation of NetBox might be missing one or more required Python packages. These packages are listed in requirements.txt and local_requirements.txt, and are normally installed as part of the installation or upgrade process. To verify installed packages, run pip freeze from the console and compare the output to the list of required packages.

**WSGI service not restarted after upgrade** - If this installation has recently been upgraded, check that the WSGI service (e.g. gunicorn or uWSGI) has been restarted. This ensures that the new code is running.

The complete exception is provided below:

```
<class 'ModuleNotFoundError'>

No module named 'geocoder'

Python version: 3.9.5
NetBox version: 3.1.6
```

```
netbox-lajolla# whoami
root
netbox-lajolla# pip install geocoder
Collecting geocoder
  Downloading geocoder-1.38.1-py2.py3-none-any.whl (98 kB)
     |████████████████████████████████| 98 kB 1.6 MB/s 
Collecting ratelim
  Downloading ratelim-0.1.6-py2.py3-none-any.whl (4.0 kB)
Collecting click
  Downloading click-8.0.3-py3-none-any.whl (97 kB)
     |████████████████████████████████| 97 kB 3.7 MB/s 
Requirement already satisfied: six in /usr/lib/python3.9/site-packages (from geocoder) (1.15.0)
Collecting future
  Downloading future-0.18.2.tar.gz (829 kB)
     |████████████████████████████████| 829 kB 11.5 MB/s 
Requirement already satisfied: requests in /usr/lib/python3.9/site-packages (from geocoder) (2.25.1)
Collecting decorator
  Downloading decorator-5.1.1-py3-none-any.whl (9.1 kB)
Requirement already satisfied: chardet<5,>=3.0.2 in /usr/lib/python3.9/site-packages (from requests->geocoder) (4.0.0)
Requirement already satisfied: idna<3.3,>=2.5 in /usr/lib/python3.9/site-packages (from requests->geocoder) (3.2)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/lib/python3.9/site-packages (from requests->geocoder) (1.26.5)
Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3.9/site-packages (from requests->geocoder) (2020.12.5)
Using legacy 'setup.py install' for future, since package 'wheel' is not installed.
Installing collected packages: decorator, ratelim, future, click, geocoder
    Running setup.py install for future ... done
Successfully installed click-8.0.3 decorator-5.1.1 future-0.18.2 geocoder-1.38.1 ratelim-0.1.6
```

```
# site_address.py

# Make sure to add `geocoder` to your `local_requirements.txt` and make sure it is installed in your Python venv.

import geocoder
from extras.reports import Report
from dcim.models import Site

class checkSiteAddress(Report):
    description = "Check if site has a physical address and/or geolocation information"

    def test_site_address(self):
        for site in Site.objects.all():
            if site.physical_address:
                self.log_success(site)
            else:
                self.log_failure(site, site.name)

    def test_site_geo(self):
        for site in Site.objects.all():
            if site.latitude and site.longitude:
                self.log_success(site)
            else:
                if site.physical_address:
                    g = geocoder.osm(site.physical_address)
                    if g:
                        self.log_warning(site, f'Missing geo location - possible ({round(g.x,6)}, {round(g.y,6)})')
                    else:
                        self.log_warning(site, f'Missing geo location ({site.latitude}, {site.longitude})')    
                else:    
                    self.log_failure(site, f'Missing geo location ({site.latitude}, {site.longitude})')
```

**Comment by candlerb** [comment]
You will need to install geocoder inside netbox's virtualenv, since that's the environment that the report will run it.

Best way is to create `/opt/netbox/local_requirements.txt` containing just the word `geocoder`, which will guarantee it's always available in future upgrades.

Then for the first time, either install into the venv as below:

```
cd /opt/netbox
. venv/bin/activate
pip install -r local_requirements.txt
```

or just run `upgrade.sh` as you would for a netbox install or upgrade.
